### PR TITLE
#8213: Implement Blob variant of WebSocket.send

### PR DESCRIPTION
**Pull request by nikkisquared**
I'm working on resolving https://github.com/servo/servo/issues/8213 as per the spec online and feedback in the servo channel. Note that currently I cannot build (and thus test) my code, so this is a bit of a rough first draft. I'd still like feedback on my progress, and I hope that there is another way for my code to be tested.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8218)

### CODE REVIEW

File: components/script/dom/websocket.rs [hunk 1]
```diff
@@ -333,6 +333,31 @@ impl WebSocketMethods for WebSocket {
...
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-websocket-send
+    fn Send(&self, data: Blob) -> Fallible<()> {
+
+        let bufferedAmount = self.buffered_amount.get() + (data.size as u32)
+
+        self.Send_Impl(bufferedAmount)
+
+        let _ = data.send(my_sender.lock().unwrap().send_message)
```
Comment (dzbarsky) [inline components/script/dom/websocket.rs]: You're not calling send_message here

**Comment by dzbarsky** [comment]
Welcome! This is a good start, but it will be pretty difficult to contribute if you're not able to compile your changes. What's preventing you from compiling? Maybe it's something we can help out with.

**Comment by nikkisquared** [comment]
@dzbarsky Thanks! I've just been having problems building servo, I don't know why as other people haven't been able to recreate it. I was able to build servo over the weekend, but after pulling changes to the repo today after I made my fork, I'm unable to build it again. I've had a whole string of different errors.

This is what happens when I build it with `./mach build -d -j1`: https://gist.github.com/nikkisquared/4f767e1a32fbd8137ab5
This is what happens when I build it with `./mach build -r -j1`: https://gist.github.com/nikkisquared/ba755bd49a8ae95536d4

File: components/script/dom/websocket.rs (hunk 1 above)
Comment (nikkisquared) [inline components/script/dom/websocket.rs]: @dzbarsky I wasn't and I'm still not sure how to use the send function for Blob. Looking at the API my best guess was it needs to be sent a function that it then calls https://github.com/servo/servo/blob/master/components/script/dom/blob.rs#L73. How should I be using it instead?

**Comment by eefriedman** [comment]
Those errors look like your build folder is corrupted somehow... maybe try make clean?

**Comment by eefriedman** [comment]
Correction: that's actually `./mach clean` for Servo.

**Comment by nikkisquared** [comment]
@eefriedman I usually run `./mach clean`, but I don't see any reason to not give it another go. I'll try it again.

File: components/script/dom/websocket.rs [hunk 2]
```diff
@@ -353,9 +378,10 @@ impl WebSocketMethods for WebSocket {
         let mut other_sender = self.sender.borrow_mut();
         let my_sender = other_sender.as_mut().unwrap();
 
-        self.buffered_amount.set(self.buffered_amount.get() + (data.0.as_bytes().len() as u32));
+        self.buffered_amount.set(bufferedAmount);
 
-        let _ = my_sender.lock().unwrap().send_message(Message::Text(data.0));
+        /*Previously, the message was sent before this section
+          will this make a difference?*/
```
Comment (jdm) [inline components/script/dom/websocket.rs]: It will not make a difference.

File: components/script/dom/websocket.rs (hunk 1 above)
Comment (jdm) [inline components/script/dom/websocket.rs]: The way http://mxr.mozilla.org/servo/source/components/script/dom/filereader.rs#357 works is it creates a channel (ie. a pair of Sender/Receiver values) that are used to extract the blob's contents. The blob sends its contents using the Sender, and whenever the consumer wants to get the actual value, it uses the receiver (see http://mxr.mozilla.org/servo/source/components/script/dom/filereader.rs#418). I recommend calling recv immediately here.

File: components/script/dom/websocket.rs [hunk 3]
```diff
@@ -353,9 +378,10 @@ impl WebSocketMethods for WebSocket {
         let mut other_sender = self.sender.borrow_mut();
         let my_sender = other_sender.as_mut().unwrap();
 
-        self.buffered_amount.set(self.buffered_amount.get() + (data.0.as_bytes().len() as u32));
+        self.buffered_amount.set(bufferedAmount);
```
Comment (jdm) [inline components/script/dom/websocket.rs]: I would suggest leaving the self.buffered_amount.get() in this calculation, and making callers pass in only the new amount to increase it by.

**Comment by nikkisquared** [comment]
@eefriedman building worked this time, and servo is building every time now. I must've forgotten to run during one of my build attempts `./mach clean`, thanks for getting me to try it again!

**Comment by nikkisquared** [comment]
I've updated it to follow what people asked for on here and irc. I've built it successfully and run tests, and the only issue tests run into now is that three of them just outright "CRASH": https://gist.github.com/nikkisquared/ca062070f531ffefeb6a. I haven't been able to figure why.

File: components/script/dom/websocket.rs [hunk 4]
```diff
@@ -333,41 +373,30 @@ impl WebSocketMethods for WebSocket {
...
-        }
 
-        /*TODO: This is not up to spec see http://html.spec.whatwg.org/multipage/comms.html search for
-                "If argument is a string"
-          TODO: Need to buffer data
-          TODO: The send function needs to flag when full by using the following
-          self.full.set(true). This needs to be done when the buffer is full
-        */
+        let data_byte_len = data.0.as_bytes().len() as u32;
+        let _ = self.Send_Impl(data_byte_len);
```
Comment (eefriedman) [inline components/script/dom/websocket.rs]: I think this should be `try!(self.Send_Impl(data_byte_len));`?  Same for `Send_`.

**Comment by nikkisquared** [comment]
@eefriedman your suggestion resolved the crashes, thanks!

**Comment by eefriedman** [comment]
This patch doesn't include any test changes; every new API needs to have a test.  (If there's an existing test in the testsuite, you can just update the .ini file appropriately.)


---

Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/websocket.rs, line 337 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8218#-K1gcmeiRGsu_sKB5SpS)** ([raw file](https://github.com/servo/servo/blob/6d2ca6b099a5d1fda5a5b3689b5b07cd19be52a4/components/script/dom/websocket.rs#L337)):</sup>
Truncating the length of the string is no good.  Fortunately, the standard gives us an out: "if the data cannot be sent, e.g. because it would need to be buffered but the buffer is full, the user agent must flag the WebSocket as full and then close the WebSocket connection."


---

<sup>**[components/script/dom/websocket.rs, line 378 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8218#-K1geFTqMBi5MHU0QdIo)** ([raw file](https://github.com/servo/servo/blob/dc8a405f244a374073f315eadca9bcba8e7cce49/components/script/dom/websocket.rs#L378)):</sup>
The `let _ =` bit here is unnecessary; the type of the expression is `()`.


---

<sup>**[components/script/dom/websocket.rs, line 394 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8218#-K1gdx0cTqW3W5f8a38H)** ([raw file](https://github.com/servo/servo/blob/dc8a405f244a374073f315eadca9bcba8e7cce49/components/script/dom/websocket.rs#L394)):</sup>
Same here: we can fail the connection if the blob is too large.


---

<sup>**[components/script/dom/websocket.rs, line 395 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8218#-K1geQWo3U_xndTW_tZk)** ([raw file](https://github.com/servo/servo/blob/dc8a405f244a374073f315eadca9bcba8e7cce49/components/script/dom/websocket.rs#L395)):</sup>
Same here: unnecessary `let _`.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8218)

**Comment by jdm** [comment]
Curiously, there are a lot of tests for WebSocket + Blob that never test sending Blobs, only receiving them. This is a weird choice. I recommend modifying `tests/wpt/web-platform-tests/websockets/Send-binary-blob.html` to use
```js
for (var i = 0; i < 65000; i++)
    data.push(String.fromCharCode(0));
data = new Blob(data);
```
instead of `data = new ArrayBuffer(datasize);`
and running the resulting test.

**Comment by nikkisquared** [comment]
I have a few questions about how to proceed on this. I think I'm doing what @eefriedman asked properly, so if you can review that I'd appreciate it.

1) @jdm I've updated that test, and when I run all of the tests it says "NOTRUN expected FAIL" for it. Where do I update the expected status of the test? And I assume I should continue updating all of the "send-blob" tests to actually use Blobs in a similar fashion.

2) I've changed it so that if the WebSocket closes, it'll delay returning until updating the buffer. I've taking a pretty basic approach to handling this. If the buffer becomes full, the WebSocket will attempt to close again. But I see nothing in the spec about updating the buffer after closing- only closing after (trying to) update the buffer. I added this only because of a note saying "TODO: Update bufferedAmount."

3) When I Close the WebSocket, what should I give for the error code and reason? The spec doesn't say what to give for either. And should I have different error messages for when the data being sent is too large for the buffer period (bigger than the maximum u32), and when the buffer becomes full from normal usage?

4) When I flag the buffer as full, should I set it to maximum capacity as well? I feel like the answer is no, but I can imagine somebody seeing the buffer flagged as full and checking the buffer, which is not actually full.

5) The contradiction I thought the HTML spec wrote itself into seems to be resolved by saying to close the WebSocket connection if the buffer is full. To be clear, the issue I saw was that the size of a Blob is u64, but the size of a buffer is u32, which doesn't add up. The resolution seems to be that if a Blob cannot fit into the buffer, it's too big to send no matter what. Maybe instead of the original issue I had planned, I can raise an issue asking for this to be pointed out clearly.

**Comment by eefriedman** [comment]
1. Take a look in `tests/wpt/metadata/websockets/`.
2. "Any invocation of this method with a Blob argument that does not throw an exception must increase the bufferedAmount attribute by the size of the Blob object's raw data, in bytes."  Calling `send()` on a closed connection doesn't throw an exception, so we need to update bufferedAmount, even though it's impossible to actually send data on the connection.
3. None; wrong kind of close.  The spec says "close the WebSocket connection", as in https://tools.ietf.org/html/rfc6455#section-7.1.1 , not send a closing handshake.
4. Saturating probably makes the most sense.  Note that you don't actually need a Blob with size greater than 2^32 to run into this issue: you can call send a bunch of times with a smaller amount of data and end up in the same situation.
5. The spec definitely needs to be clarified as to what should happen if bufferedAmount overflows.


---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8218)

**Comment by jdm** [comment]
I would be curious to see the full output of that test. I'm not sure why any part of it wouldn't be run.

**Comment by nikkisquared** [comment]
@jdm I checked the logs and it says "data.push is not a function", so I need a different inner loop for https://github.com/servo/servo/pull/8218#issuecomment-151837940.

@eefriedman thanks for the detailed response as always!

1) Do the ini files here only list tests that are expected to do something other than pass? I don't see the first test in Send-binary-blob listed here.

2) Alright then. Does the code for returning early after buffering seem fine? For some reason I get a compiler warning that I'm not using `return_after_buffer` even though there's an if statement using it, so it sounds like I could improve it.

3) Alright, so the Close() function is for something else. What do I call to close the WebSocket Connection? My next guess is that it'd be https://github.com/nikkisquared/servo/blob/master/components/script/dom/websocket.rs#L518, but I don't know what most of these do.

4) I've set the buffered amount to the max u32 value in that case then. Saturating a u64 value that's too large, or a value that'd push the buffer over, would result in the same thing, right? I know the issue can happen by sending smaller data points over and over, I was wondering if that situation would count as a different error than being sent a message that is too large to send even on an empty buffer. But if the function I need for point 3 doesn't let me specify the exact error like Close() does, there's no reason to try to make a distinction.

5) I'll open an issue at the whatwg/html repo asking about this.

**Comment by eefriedman** [comment]
> 1) Do the ini files here only list tests that are expected to do something other than pass? I don't see the first test in Send-binary-blob listed here.

Yes.

> For some reason I get a compiler warning that I'm not using return_after_buffer

`let return_after_buffer = true;` creates a new binding; it doesn't mutate the existing variable.

> What do I call to close the WebSocket Connection?

https://github.com/nikkisquared/servo/blob/master/components/script/dom/websocket.rs#L250 ?

> Saturating a u64 value that's too large, or a value that'd push the buffer over, would result in the same thing, right?

Yes.

**Comment by jdm** [comment]
Oh, my mistake. It should be `data += String.fromCharCode(0);` instead. That will teach me to retype from memory instead of copying and pasting.

**Comment by nikkisquared** [comment]
@eefriedman Send_Impl now uses the CloseTask function. I re-arranged creating the message channel and address url to the top of Send_Impl so both the new CloseTask call and BufferedAmountTask can access them without redundancy. I don't know how this might affect efficiency, since the variables are created every time data is sent, regardless of being used. Should I make a function for returning them, store them in the WebSocket, or something else? From what I understand if the values are easily accessible it might impact security, so I want to be sure I'm doing the best thing.

@jdm Thanks! I'm going to go ahead and update the other "blob" tests in a similar way and make sure they pass.

**Comment by nikkisquared** [comment]
There was only one other pre-existing test dealing with Blobs. I assume more testing is required than just those two, but I don't know where to start.

**Comment by eefriedman** [comment]
Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions.

---

<sup>**[components/script/dom/websocket.rs, line 378 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8218#-K1geFTqMBi5MHU0QdIo)** ([raw file](https://github.com/servo/servo/blob/dc8a405f244a374073f315eadca9bcba8e7cce49/components/script/dom/websocket.rs#L378)):</sup>
The let _ = bit here is unnecessary; the type of the expression is ().


---

<sup>**[components/script/dom/websocket.rs, line 303 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8218#-K1pxbEPgcOt_5M3flzV)** ([raw file](https://github.com/servo/servo/blob/1bc0bcefb192e38c564dbf21daa0e3a3806db948/components/script/dom/websocket.rs#L303)):</sup>
`Trusted::new(global.r().get_cx(), self, chan`?

It probably makes sense to move these variables after the ready_state check.

The performance doesn't really matter because the cost of calling send on a closed socket isn't usually relevant.


---

<sup>**[components/script/dom/websocket.rs, line 325 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8218#-K1q2VJBDt1-e7kz4D7T)** ([raw file](https://github.com/servo/servo/blob/1bc0bcefb192e38c564dbf21daa0e3a3806db948/components/script/dom/websocket.rs#L325)):</sup>
This closes the connection from the perspective of the script... but it doesn't actually close the TCP connection.


---

<sup>**[components/script/dom/websocket.rs, line 334 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8218#-K1pzjeb7st4VUJ0WzK4)** ([raw file](https://github.com/servo/servo/blob/1bc0bcefb192e38c564dbf21daa0e3a3806db948/components/script/dom/websocket.rs#L334)):</sup>
I think Send_Impl actually needs to return a `Fallible<bool>`, which returns false if don't actually want to send data; otherwise we'll try to send data on a connection which should be closed.

Also, probably need to check self.full somewhere.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8218)

**Comment by jdm** [comment]
I did a bit more investigating for the tests; we'll need to use `data = new Blob([data])`, since our Blob constructor in Servo isn't spec-compliant right now. The result should end up the same in Servo, however.

**Comment by jdm** [comment]
With regard to closing the TCP connection, I don't see a convenient mechanism to do so in http://doc.servo.org/websocket/index.html . Possibly it requires the sender and receiver to go out of scope, but I don't think it's worth spending a lot of time to try to get that right in this PR. I suggest we call `WebSocket::Close` directly (as in an earlier commit) with a TODO comment (and issue filed) to implement a proper immediate TCP close later.

**Comment by nikkisquared** [comment]
@jdm I updated the two tests I've worked on before to have that change. Are there other tests I should be updating likewise?

@eefriedman I've removed the unnecessary `let _ = ...` statements. I must have accidentally added them back in after removing them the first time! I've also moved global, chan, and address declarations. And Send_Impl now returns Falliable<bool>.

As for closing the WebSocket properly, there's what @jdm has said already. Is there anything I should give for the error code or reason? Like I said before, I'd like having a reason declared to help with debugging purposes.

> Also, probably need to check self.full somewhere.
This also seems out of scope for this PR, and jdm told me in irc that there's no actual buffer yet, so there isn't going to be an issue with it actually being full. I did notice that the send_close function made in WebSocket.Close has a TODO comment for checking if the buffer is full. That seems like the place where it'd be done, so maybe I could extend that comment or something.

**Comment by eefriedman** [comment]
> Is there anything I should give for the error code or reason?

Doesn't really matter; it'll go away eventually.  (https://github.com/cyderize/rust-websocket/pull/52 etc.)

> And Send_Impl now returns `Falliable<bool>`.

That isn't really useful unless you actually check the result.  Also, the `return_after_buffer` case should return false.

About `self.full`, whatever, we can deal with it in a followup.

**Comment by nikkisquared** [comment]
> That isn't really useful unless you actually check the result. Also, the return_after_buffer case should return false.

@eefriedman I thought I was missing something! I'm doing both of those now.

What more should I do for my PR?

**Comment by eefriedman** [comment]
Looks good except for a few small things!  Thanks for sticking with this. :)


---

Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[components/script/dom/websocket.rs, line 313 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8218#-K2KcWgKTjJWaRZNYulG)** ([raw file](https://github.com/servo/servo/blob/1b35613b34926a7ecb1c1cfea92ce4e45e54edb2/components/script/dom/websocket.rs#L313)):</sup>
`Trusted::new(global.r().get_cx(), self, chan)`?


---

<sup>**[components/script/dom/websocket.rs, line 318 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8218#-K2KccMAGpxt9ZbagY7K)** ([raw file](https://github.com/servo/servo/blob/1b35613b34926a7ecb1c1cfea92ce4e45e54edb2/components/script/dom/websocket.rs#L318)):</sup>
Extra newline.


---

<sup>**[components/script/dom/websocket.rs, line 325 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8218#-K2Kcf-ywriW930cjIx7)** ([raw file](https://github.com/servo/servo/blob/1b35613b34926a7ecb1c1cfea92ce4e45e54edb2/components/script/dom/websocket.rs#L325)):</sup>
Extra newline.


---

<sup>**[components/script/dom/websocket.rs, line 391 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8218#-K2Kclk4g-7r7ADsPA4x)** ([raw file](https://github.com/servo/servo/blob/1b35613b34926a7ecb1c1cfea92ce4e45e54edb2/components/script/dom/websocket.rs#L391)):</sup>
Trailing whitespace; `./mach test-tidy` will catch this.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8218)

**Comment by nikkisquared** [comment]
> Looks good except for a few small things! Thanks for sticking with this. :)

Thanks! I've been enjoying my time working on this.

> Trusted::new(global.r().get_cx(), self, chan)?

Oh, right I reverted it because it gave me a compiler error to use `chan` there. I knew I forgot to mention something else. This is the error I get: https://gist.github.com/nikkisquared/495cebc6d0c9910b6063.

I fixed those extra whitespaces and a few others `./mach test-tidy` caught. I'll remember to do that in the future, it's handy!

**Comment by eefriedman** [comment]
Oh... you can use `chan.clone()`.

Also, please squash the commits.

**Comment by nikkisquared** [comment]
@eefriedman Done and done!

**Pull request merged**